docs/design-proposals/etcd/ETCD-Data-Partition-Design.md

```diff
@@ -51,17 +54,156 @@ number, watch from revision number, etc.
+created/owned by tenants, including namespaces for a tenant, pods created in one of the namespaces
+of the tenants, etc. A ETCD tenant cluster can have objects from multiple tenants. Objects for a single
+tenant can only located in one tenant cluster.
+
+#### API Changes
+
+* We added a top level Kubernetes object named '**TenantStorage**' with path '**/api/v1/tenantstorages**'.
+In addition, we added a field named '**StorageClusterId**' into '**[Tenant](../multi-tenancy/multi-tenancy-api-resource-model.md)**' object.
+  - This allows admin to view and allocate tenants to etcd clusters manually
+  - This allows api server to discover tenant clusters and decide where to create/locate objects
+
+```yaml
+apiVersion: v1
+kind: TenantStorage 
+storageClusterId: "c1"
+serviceAddress: "172.30.0.122:2379"
+metadata:
+  name: "tenant-cluster-1"
+```
+
+```yaml
+apiVersion: v1
+kind: Tenant
+spec:
+  storageClusterId: c1
+metadata:
+  name: aa 
+```
+
+```text
+$ curl http://127.0.0.1:8080/api/v1/tenantstorages
+{
+  "kind": "TenantStorageList",
+  "apiVersion": "v1",
+  "metadata": {
+    "selfLink": "/api/v1/tenantstorages",
+    "resourceVersion": "833715344073826304"
+  },
+  "items": [
+    {
+      "metadata": {
+        "name": "tenant-cluster-1",
+        "selfLink": "/api/v1/tenantstorages/tenant-cluster-1",
+        "uid": "9381a173-b138-4e3d-bf64-98e8dd7043cb",
+        "hashKey": 1006587480031365457,
+        "resourceVersion": "833714719937273857",
+        "creationTimestamp": "2020-05-22T21:55:42Z",
+        "annotations": {
+          "kubectl.kubernetes.io/last-applied-configuration": "{\"apiVersion\":\"v1\",\"kind\":\"TenantStorage\",\"metadata\":{\"annotations\":{},\"name\":\"tenant-cluster-1\"},\"serviceAddress\":\"172.30.0.122:2379\",\"storageClusterId\":\"c1\"}\n"
+        }
+      },
+      "storageClusterId": "c1",
+      "serviceAddress": "172.30.0.122:2379"
+    }
+  ]
+}
+```
+
+```text
+$ curl http://127.0.0.1:8080/api/v1/tenants/aa
+{
+  "kind": "Tenant",
+  "apiVersion": "v1",
+  "metadata": {
+    "name": "aa",
+    "selfLink": "/api/v1/tenants/aa",
+    "uid": "150adc36-885f-46c3-9db4-c6f9876fcf01",
+    "hashKey": 2305167293877862255,
+    "resourceVersion": "833715489820647426",
+    "creationTimestamp": "2020-05-22T22:21:41Z"
+  },
+  "spec": {
+    "finalizers": [
+      "kubernetes"
+    ],
+    "storageClusterId": "c1"
+  },
+  "status": {
+    "phase": "Active"
+  }
+}
+```
+
+#### API Structs
+
+Using storage cluster identified by tenant storage, Api servers can have automatically connect
+to ETCD cluster(s) that hosting specific tenant data.
+  
+Here are the TenantStorage data struct introduced into Arktos:
+
+```text
+type TenantStorage struct {
+	metav1.TypeMeta
+	metav1.ObjectMeta
+
+	// A string that specifies the storage object identity
+	StorageClusterId string
+
+	// A string that specifies the backend storage server address
+	ServiceAddress string
```

Comment:
I suppose it it the etcd address. Therefore the  cluster id shall be in pair with etcd, right?

Reply:
> I suppose it is an etcd address. Therefore the cluster id shall be in pair with etcd, right?

It is an etcd address. They are 1:1 by now. Might need to change to 1:m as we might need to follow the convention of original implementation that allows multiple addresses per cluster.